In [1]:
import boto3
import pandas as pd
from io import StringIO

# Define your bucket name and file key (file path in S3)
BUCKET_NAME = "blue-blood-data"
FILE_KEY = "BigQuery Blue Blood DB Data.csv"  # Change to your actual file path in S3

# Create an S3 client
s3 = boto3.client("s3")

# Fetch the file from S3
response = s3.get_object(Bucket=BUCKET_NAME, Key=FILE_KEY)

# Read the CSV file into a pandas DataFrame
csv_content = response["Body"].read().decode("utf-8")
df = pd.read_csv(StringIO(csv_content))

# Print DataFrame
print(df.head())

   subject_id   prescription_start  \
0       10013  2125-10-05T00:00:00   
1       10019  2163-05-15T00:00:00   
2       10027  2190-07-14T00:00:00   
3       10040  2147-02-24T00:00:00   
4       10042  2147-02-07T00:00:00   

                                  prescription_drugs  \
0  Dobutamine, Vasopressin, Vasopressin, Ceftriax...   
1  Sodium Chloride , Sodium Chloride , Norepineph...   
2  Dextrose 50%, Amiodarone HCl, Amiodarone HCl, ...   
3  Magnesium Sulfate, Multivitamins, Bisacodyl, F...   
4  Nitroprusside Sodium, Meperidine, Midazolam HC...   

                         prescription_drug_name_poes  \
0  Azithromycin, Azithromycin, Morphine Sulfate, ...   
1  Sodium Bicarbonate, Artificial Tear Ointment, ...   
2  Dextrose 50%, Amiodarone HCl, HydrALAZINE HCl,...   
3  Magnesium Sulfate, Multivitamins, Bisacodyl, F...   
4  Meperidine, Midazolam HCl, Magnesium Sulfate, ...   

                     prescription_formulary_drug_cds  \
0  DOBPREM, VASO20I, VASO20I, CEFX1F, DOP

In [2]:
df 

subject_id   prescription_start  \
0         10013  2125-10-05T00:00:00   
1         10019  2163-05-15T00:00:00   
2         10027  2190-07-14T00:00:00   
3         10040  2147-02-24T00:00:00   
4         10042  2147-02-07T00:00:00   
..          ...                  ...   
276       42346  2160-12-21T00:00:00   
277       42346  2160-12-26T00:00:00   
278       43735  2128-11-09T00:00:00   
279       44154  2178-05-15T00:00:00   
280       44212  2123-11-25T00:00:00   

                                    prescription_drugs  \
0    Dobutamine, Vasopressin, Vasopressin, Ceftriax...   
1    Sodium Chloride , Sodium Chloride , Norepineph...   
2    Dextrose 50%, Amiodarone HCl, Amiodarone HCl, ...   
3    Magnesium Sulfate, Multivitamins, Bisacodyl, F...   
4    Nitroprusside Sodium, Meperidine, Midazolam HC...   
..                                                 ...   
276             Potassium Chloride, Potassium Chloride   
277  Morphine Sulfate, Famotidine, D5W, Heparin, Fe...   
278  Norepinephrine, Midazolam, Fentanyl Citrate, M...   
279  Vasopressin, Norepinephrine, DOBUTamine, DOPam...   
280  Midazolam, Fentanyl Citrate, Piperacillin-Tazo...   

                           prescription_drug_name_poes  \
0    Azithromycin, Azithromycin, Morphine Sulfate, ...   
1    Sodium Bicarbonate, Artificial Tear Ointment, ...   
2    Dextrose 50%, Amiodarone HCl, HydrALAZINE HCl,...   
3    Magnesium Sulfate, Multivitamins, Bisacodyl, F...   
4    Meperidine, Midazolam HCl, Magnesium Sulfate, ...   
..                                                 ...   
276             Potassium Chloride, Potassium Chloride   
277              Famotidine, Heparin, Fentanyl Citrate   
278  Midazolam, Fentanyl Citrate, Albuterol Inhaler...   
279  Calcium Gluconate, Albuterol Inhaler, Chlorhex...   
280                             Cisatracurium Besylate   

                       prescription_formulary_drug_cds  \
0    DOBPREM, VASO20I, VASO20I, CEFX1F, DOPA400PM, ...   
1    NACL23.4I, NACL23.4I, LEVO4I, SW3000, SW3000, ...   
2    DEX50SY, AMIO150I, AMIO150I, AMIO150I, HYDZ20I...   
3    MAGS1I, MVI, BISA5, FERR325, FURO40I, PANT40, ...   
4    NITP50I, MEPE50I, MIDA2I, MAGS1I, LEVO4I, VANC...   
..                                                 ...   
276                                     POTA20, KCL20P   
277  MORP100PM, PEPC20, MORPD5W, HEPA5I, FENT2I, NS...   
278  LEVO4I, MIDA100, FENT2.550, MORP100PM, PROAIR8...   
279  VASO20I, LEVO4I, DOBPREM, DOPA400D, LEVO750PM,...   
280  MIDA50I, FENT2.5I, ZOSY2.25I, CISA200I, CISA20...   

                                     prescription_gsns  \
0    21502, 6612, 6612, 9156, 4975, 4975, 6522, 166...   
1    1220, 1220, 4939, 9825, 9825, 9825, 9825, 9825...   
2    1989, 50676, 50676, 50676, 283, 283, 8205, 197...   
3    1417, 2524, 2947, 1645, 8205, 27462, 6549, 300...   
4    19652, 4051, 3779, 1412, 4939, 9329, 23159, 15...   
..                                                 ...   
276                                        22346, 1262   
277              59940, 11677, 6549, 41384, 1210, 1723   
278  28633, 3780, 41385, 59940, 28090, 59081, 4817,...   
279  6612, 28633, 21502, 4975, 46770, 9588, 63951, ...   
280  3780, 41385, 21185, 24613, 24613, 24613, 24613...   

                                     prescription_ndcs  \
0    74234632, 54569391600, 54569391600, 4200278, 7...   
1    517293025, 517293025, 74144304, 74797308, 7479...   
2    409490234, 55390005710, 55390005710, 553900057...   
3    517260225, 55499120401, 51079000620, 510790116...   
4    74302401, 54354563, 74258702, 74407532, 741443...   
..                                                 ...   
276                             58177020211, 456066270   
277  409606211, 51079096620, 0, 63323026201, 100190...   
278  703115303, 10019002710, 10019003783, 409606211...   
279  63323030201, 703115303, 409234632, 338100702, ...   
280  10019002710, 10019003783, 206885216, 74438220,...   

                           prescription

In [3]:
import pandas as pd
import numpy as np

# Print values of columns that start with 'pre_' or 'post_'
'''for col in df.columns:
    if col.startswith("pre_") or col.startswith("post_"):
        print(f"Column: {col}")
        print(df[col])'''
for col in df.select_dtypes(include=[np.number]).columns:
    if (df[col] < 0).any():
        print(f"Column '{col}' contains negative values.", df[col].values.min())


Column 'pre_baseexcess' contains negative values. -16.0
Column 'post_baseexcess' contains negative values. -17.0


In [4]:
import pandas as pd
import numpy as np

# Create a new DataFrame with columns starting with 'pre_' or 'post_'
filtered_df = df.loc[:, df.columns.str.startswith(('pre_', 'post_'))]

# Replace NaN values with -200 in the filtered DataFrame
filtered_df = filtered_df.fillna(-200)

# Show the resulting DataFrame
print(filtered_df)


           pre_charttime  pre_ph  pre_pco2  pre_po2  pre_bicarbonate  \
0    2125-10-04T23:59:00    7.30      63.0     60.0           -200.0   
1    2163-05-14T23:11:00    7.29      35.0     69.0           -200.0   
2    2190-07-13T23:37:00    7.35      35.0    129.0           -200.0   
3    2147-02-23T13:28:00    7.40      37.0    232.0           -200.0   
4    2147-02-06T13:45:00    7.46      37.0     81.0           -200.0   
..                   ...     ...       ...      ...              ...   
276  2160-12-17T03:28:00    7.38      68.0     66.0           -200.0   
277  2160-12-17T03:28:00    7.38      68.0     66.0           -200.0   
278  2128-11-08T12:38:00    7.31      43.0     80.0           -200.0   
279  2178-05-14T21:59:00    7.26      47.0     51.0           -200.0   
280  2123-11-24T21:56:00    7.24      48.0     76.0           -200.0   

     pre_baseexcess  pre_totalco2  pre_hematocrit  pre_hemoglobin  pre_sodium  \
0               2.0          32.0          -200.0     

In [5]:
import pandas as pd
import numpy as np

# Check and print columns that contain NaN values
for col in df.columns:
    if df[col].isna().any():  # Check if there are any NaN values in the column
        print(f"Column: {col}")
        print(df[col])

Column: prescription_drug_name_poes
0      Azithromycin, Azithromycin, Morphine Sulfate, ...
1      Sodium Bicarbonate, Artificial Tear Ointment, ...
2      Dextrose 50%, Amiodarone HCl, HydrALAZINE HCl,...
3      Magnesium Sulfate, Multivitamins, Bisacodyl, F...
4      Meperidine, Midazolam HCl, Magnesium Sulfate, ...
                             ...                        
276               Potassium Chloride, Potassium Chloride
277                Famotidine, Heparin, Fentanyl Citrate
278    Midazolam, Fentanyl Citrate, Albuterol Inhaler...
279    Calcium Gluconate, Albuterol Inhaler, Chlorhex...
280                               Cisatracurium Besylate
Name: prescription_drug_name_poes, Length: 281, dtype: object
Column: prescription_gsns
0      21502, 6612, 6612, 9156, 4975, 4975, 6522, 166...
1      1220, 1220, 4939, 9825, 9825, 9825, 9825, 9825...
2      1989, 50676, 50676, 50676, 283, 283, 8205, 197...
3      1417, 2524, 2947, 1645, 8205, 27462, 6549, 300...
4      19652, 4051, 3

In [6]:
import pandas as pd
import numpy as np

# Select columns that start with 'pre_' or 'post_'
selected_columns = df.loc[:, df.columns.str.startswith(('pre_', 'post_'))]

# Ensure that we only work with numeric columns
numeric_selected_columns = selected_columns.select_dtypes(include=[np.number])

# Find the maximum value across all numeric values in the selected columns (ignoring NaN values)
max_value = numeric_selected_columns.max().max()

# Now, divide only the numeric values in the selected columns by the max_value
df.loc[:, numeric_selected_columns.columns] = numeric_selected_columns / max_value

# Show the resulting DataFrame
print(df)


     subject_id   prescription_start  \
0         10013  2125-10-05T00:00:00   
1         10019  2163-05-15T00:00:00   
2         10027  2190-07-14T00:00:00   
3         10040  2147-02-24T00:00:00   
4         10042  2147-02-07T00:00:00   
..          ...                  ...   
276       42346  2160-12-21T00:00:00   
277       42346  2160-12-26T00:00:00   
278       43735  2128-11-09T00:00:00   
279       44154  2178-05-15T00:00:00   
280       44212  2123-11-25T00:00:00   

                                    prescription_drugs  \
0    Dobutamine, Vasopressin, Vasopressin, Ceftriax...   
1    Sodium Chloride , Sodium Chloride , Norepineph...   
2    Dextrose 50%, Amiodarone HCl, Amiodarone HCl, ...   
3    Magnesium Sulfate, Multivitamins, Bisacodyl, F...   
4    Nitroprusside Sodium, Meperidine, Midazolam HC...   
..                                                 ...   
276             Potassium Chloride, Potassium Chloride   
277  Morphine Sulfate, Famotidine, D5W, Heparin, Fe... 

In [8]:
filtered_df

pre_charttime  pre_ph  pre_pco2  pre_po2  pre_bicarbonate  \
0    2125-10-04T23:59:00    7.30      63.0     60.0           -200.0   
1    2163-05-14T23:11:00    7.29      35.0     69.0           -200.0   
2    2190-07-13T23:37:00    7.35      35.0    129.0           -200.0   
3    2147-02-23T13:28:00    7.40      37.0    232.0           -200.0   
4    2147-02-06T13:45:00    7.46      37.0     81.0           -200.0   
..                   ...     ...       ...      ...              ...   
276  2160-12-17T03:28:00    7.38      68.0     66.0           -200.0   
277  2160-12-17T03:28:00    7.38      68.0     66.0           -200.0   
278  2128-11-08T12:38:00    7.31      43.0     80.0           -200.0   
279  2178-05-14T21:59:00    7.26      47.0     51.0           -200.0   
280  2123-11-24T21:56:00    7.24      48.0     76.0           -200.0   

     pre_baseexcess  pre_totalco2  pre_hematocrit  pre_hemoglobin  pre_sodium  \
0               2.0          32.0          -200.0          -200.0      -200.0   
1              -8.0          18.0          -200.0          -200.0      -200.0   
2              -5.0          20.0          -200.0          -200.0      -200.0   
3               0.0          24.0          -200.0          -200.0      -200.0   
4               2.0          27.0          -200.0          -200.0      -200.0   
..              ...           ...             ...             ...         ...   
276            11.0          42.0          -200.0          -200.0      -200.0   
277            11.0          42.0          -200.0          -200.0      -200.0   
278            -4.0          23.0          -200.0          -200.0      -200.0   
279            -5.0          22.0          -200.0          -200.0      -200.0   
280            -6.0          22.0          -200.0          -200.0      -200.0   

     ...  post_fio2_chartevents  post_aado2_calc  post_pao2fio2  \
0    ...              60.000002       231.550017     191.666659   
1    ...            -200.000000      -200.000000    -200.000000   
2    ...              40.000001       113.200004     329.999995   
3    ...              40.000001        74.700000     395.000000   
4    ...            -200.000000      -200.000000    -200.000000   
..   ...                    ...              ...            ...   
276  ...            -200.000000      -200.000000    -200.000000   
277  ...            -200.000000      -200.000000    -200.000000   
278  ...              50.000000       379.650000     171.250000   
279  ...             100.000000       618.000000      60.000000   
280  ...              80.000000       343.900000     198.750000   

     post_temperature  post_fio2  post_aado2  post_carboxyhemoglobin  \
0              -200.0     -200.0      -200.0                  -200.0   
1              -200.0     -200.0      -200.0                  -200.0   
2              -200.0     -200.0      -200.0                  -200.0   
3                37.9       40.0      -200.0                  -200.0   
4              -200.0     -200.0      -200.0                  -200.0   
..                ...        ...         ...                     ...   
276            -200.0     -200.0      -200.0                  -200.0   
277            -200.0     -200.0      -200.0                  -200.0   
278            -200.0       80.0       395.0                  -200.0   
279            -200.0     -200.0      -200.0                  -200.0   
280            -200.0     -200.0      -200.0                  -200.0   

     post_methemoglobin  post_calcium  post_intubated  
0                -200.0       -200.00          -200.0  
1                -200.0          0.92          -200.0  
2                -200.0       -200.00          -200.0  
3                -200.0       -200.00          -200.0  
4                -200.0          1.09          -200.0  
..                  ...           ...             ...  
276              -200.0       -200.00          -200.0  
277              -200.0       -200.00          -200.0  

In [20]:
filtered_df = filtered_df.drop(columns=['pre_charttime', 'post_charttime'],axis=1)

In [22]:
from sklearn.preprocessing import MinMaxScaler

scaler = MinMaxScaler()

ndArray = scaler.fit_transform(filtered_df)

In [25]:
df_scaled = pd.DataFrame(ndArray, columns = filtered_df.columns)

df_scaled

pre_ph  pre_pco2   pre_po2  pre_bicarbonate  pre_baseexcess  \
0    0.500000  0.741935  0.023018              0.0        0.545455   
1    0.479167  0.290323  0.046036              0.0        0.242424   
2    0.604167  0.290323  0.199488              0.0        0.333333   
3    0.708333  0.322581  0.462916              0.0        0.484848   
4    0.833333  0.322581  0.076726              0.0        0.545455   
..        ...       ...       ...              ...             ...   
276  0.666667  0.822581  0.038363              0.0        0.818182   
277  0.666667  0.822581  0.038363              0.0        0.818182   
278  0.520833  0.419355  0.074169              0.0        0.363636   
279  0.416667  0.483871  0.000000              0.0        0.333333   
280  0.375000  0.500000  0.063939              0.0        0.303030   

     pre_totalco2  pre_hematocrit  pre_hemoglobin  pre_sodium  pre_potassium  \
0        0.594595             0.0             0.0         0.0       0.000000   
1        0.216216             0.0             0.0         0.0       0.000000   
2        0.270270             0.0             0.0         0.0       0.987846   
3        0.378378             0.0             0.0         0.0       0.000000   
4        0.459459             0.0             0.0         0.0       0.000000   
..            ...             ...             ...         ...            ...   
276      0.864865             0.0             0.0         0.0       0.989791   
277      0.864865             0.0             0.0         0.0       0.989791   
278      0.351351             0.0             0.0         0.0       0.000000   
279      0.324324             0.0             0.0         0.0       0.000000   
280      0.324324             0.0             0.0         0.0       0.000000   

     ...  post_fio2_chartevents  post_aado2_calc  post_pao2fio2  \
0    ...               0.866667         0.527567       0.548553   
1    ...               0.000000         0.000000       0.000000   
2    ...               0.800000         0.382885       0.742297   
3    ...               0.800000         0.335819       0.833333   
4    ...               0.000000         0.000000       0.000000   
..   ...                    ...              ...            ...   
276  ...               0.000000         0.000000       0.000000   
277  ...               0.000000         0.000000       0.000000   
278  ...               0.833333         0.708619       0.519958   
279  ...               1.000000         1.000000       0.364146   
280  ...               0.933333         0.664914       0.558473   

     post_temperature  post_fio2  post_aado2  post_carboxyhemoglobin  \
0            0.000000   0.000000    0.000000                     0.0   
1            0.000000   0.000000    0.000000                     0.0   
2            0.000000   0.000000    0.000000                     0.0   
3            0.995814   0.800000    0.000000                     0.0   
4            0.000000   0.000000    0.000000                     0.0   
..                ...        ...         ...                     ...   
276          0.000000   0.000000    0.000000                     0.0   
277          0.000000   0.000000    0.000000                     0.0   
278          0.000000   0.933333    0.713429                     0.0   
279          0.000000   0.000000    0.000000                     0.0   
280          0.000000   0.000000    0.000000                     0.0   

     post_methemoglobin  post_calcium  post_intubated  
0                   0.0      0.000000             0.0  
1                   0.0      0.998559             0.0  
2                   0.0      0.000000             0.0  
3                   0.0      0.000000             0.0  
4                   0.0      0.999404             0.0  
..                  ...           ...             ...  
276                 0.0      0.000000             0.0  
277                 0.0      0.000000             0.0  
278                 0.0      0.

##Variables to remember
df contains the data before preprocessing
filtered_data contains -200 in the place of NULL values
df_scaled has the scaled data

281 x 50 columns